Sources/SwiftParser/Parser.swift

```diff
@@ -629,3 +658,69 @@ extension Parser {
     )
   }
 }
+
+// MARK: Incremental Parsing
+extension Parser {
+  mutating func loadCurrentSyntaxNodeFromCache(for kind: SyntaxKind) -> Syntax? {
+    guard parseLookup != nil else {
+      return nil
+    }
+
+    let currentOffset = self.lexemes.getOffsetToStart(self.currentToken)
+    if let node = parseLookup!.lookUp(currentOffset, kind: kind) {
+      self.lexemes.advance(by: node.byteSize, currentToken: &self.currentToken)
+      return node
+    }
+
+    return nil
+  }
+
+  mutating func registerNodeForIncrementalParse(node: RawSyntax, startToken: Lexer.Lexeme) {
+    lookaheadRanges.registerNodeForIncrementalParse(
+      node: node,
+      lookaheadLength: lexemes.lookaheadTracker.pointee.furthestOffset - self.lexemes.getOffsetToStart(startToken)
+    )
+  }
+}
+
+/// Record the furthest offset to `sourceBufferStart` that is reached by  `Parser.peek()` or `Parser.Lookahead` in ``Lexer/LexemeSequence``
```

Comment:
Depending on whether `LexemeSequence` or `Parser`/`Lookahead` is responsible fro calling `recordFurthestOffset`